gobblin-modules/gobblin-kafka-common/src/test/java/org/apache/gobblin/metrics/reporter/KafkaAvroEventKeyValueReporterTest.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.gobblin.metrics.MetricContext;
+import org.apache.gobblin.metrics.kafka.KafkaAvroEventKeyValueReporter;
+import org.apache.gobblin.metrics.kafka.KafkaEventReporter;
+import org.apache.gobblin.metrics.kafka.Pusher;
+import org.apache.gobblin.metrics.reporter.util.EventUtils;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Properties;
+
+public class KafkaAvroEventKeyValueReporterTest extends KafkaAvroEventReporterTest {
+  Properties properties;
+
+  @BeforeClass
+  public void setup() {
+    properties = new Properties();
+    properties.setProperty(KafkaAvroEventKeyValueReporter.keyPropertyName, "k1,k2,k3");
+  }
+
+  @Override
+  public KafkaEventReporter.Builder<? extends KafkaEventReporter.Builder> getBuilder(MetricContext context,
+                                                                                     Pusher pusher) {
+    return new KafkaAvroEventKeyValueReporter.BuilderImpl(context, properties).withKafkaPusher(pusher);
+  }
+
+  private Pair<String, GobblinTrackingEvent> nextKVEvent(Iterator<Pair<String, byte[]>> it) throws IOException {
+    Assert.assertTrue(it.hasNext());
+    Pair<String, byte[]> event = it.next();
+    return Pair.of(event.getKey(), EventUtils.deserializeReportFromAvroSerialization(new GobblinTrackingEvent(), event.getValue()));
+  }
+
+  @Test
+  public void testKafkaEventReporter() throws IOException {
+    MetricContext context = MetricContext.builder("context").build();
+
+    MockKafkaKeyValuePusher pusher = new MockKafkaKeyValuePusher();
+    KafkaEventReporter kafkaReporter = getBuilder(context, pusher).build("localhost:0000", "topic");
+
+    String namespace = "gobblin.metrics.test";
+    String eventName = "testEvent";
+
+    GobblinTrackingEvent event = new GobblinTrackingEvent();
+    event.setName(eventName);
+    event.setNamespace(namespace);
+    Map<String, String> metadata = Maps.newHashMap();
+    metadata.put("m1", "v1");
+    metadata.put("m2", null);
+    event.setMetadata(metadata);
+    context.submitEvent(event);
+
+    try {
+      Thread.sleep(100);
+    } catch(InterruptedException ex) {
+      Thread.currentThread().interrupt();
+    }
+
+    kafkaReporter.report();
+
+    try {
+      Thread.sleep(100);
+    } catch(InterruptedException ex) {
+      Thread.currentThread().interrupt();
+    }
+
+    Pair<String, GobblinTrackingEvent> retrievedEvent = nextKVEvent(pusher.messageIterator());
+    Assert.assertNull(retrievedEvent.getKey());
+
+    event = new GobblinTrackingEvent();
+    event.setName(eventName);
+    event.setNamespace(namespace);
+    metadata = Maps.newHashMap();
+    metadata.put("k1", "v1");
+    metadata.put("k2", "v2");
+    metadata.put("k3", "v3");
+    event.setMetadata(metadata);
+    context.submitEvent(event);
+
+    try {
```

Comment:
Why are the sleeps required?

Reply:
Not sure, but KafkaEventReporterTest also have sleep, so I guess event submission and reporting might need some time...

Reply:
That could make this flaky. Can you remove the sleeps and see what happens? I don't think there are any other threads involved here, so I'm not sure the sleeps are required.

Reply:
removed